app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/uploadsupportingdocs/UploadSupportingDocumentsController.scala

```diff
@@ -0,0 +1,137 @@
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.returns.UploadSupportingDocuments.IncompleteUploadSupportingDocuments
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.returns.{SingleDisposalDraftReturn, UploadSupportingDocuments}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.{BooleanFormatter, SessionData}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.repos.SessionStore
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.Logging
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.returns.{uploadsupportingdocs => pages}
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class UploadSupportingDocumentsController @Inject() (
+  val authenticatedAction: AuthenticatedAction,
+  val sessionDataAction: SessionDataAction,
+  val sessionStore: SessionStore,
+  val errorHandler: ErrorHandler,
+  cc: MessagesControllerComponents,
+  val config: Configuration,
+  hasSupportingDocsToUploadPage: pages.has_supporting_docs_to_upload
+)(implicit viewConfig: ViewConfig, ec: ExecutionContext)
+    extends FrontendController(cc)
+    with WithAuthAndSessionDataAction
+    with Logging
+    with SessionUpdates {
+
+  private def withSupportingDocumentChoices(
+    request: RequestWithSessionData[_]
+  )(
+    f: (
+      SessionData,
+      FillingOutReturn,
+      SingleDisposalDraftReturn,
+      UploadSupportingDocuments
+    ) => Future[Result]
+  ): Future[Result] =
+    request.sessionData.flatMap(s => s.journeyStatus.map(s -> _)) match {
+      case Some((s, r @ FillingOutReturn(_, _, _, d: SingleDisposalDraftReturn))) =>
+        d.uploadSupportingDocuments.fold[Future[Result]](
+          f(s, r, d, IncompleteUploadSupportingDocuments.empty)
+        )(f(s, r, d, _))
+      case _ => Future.successful(Redirect(controllers.routes.StartController.start()))
+    }
+
+  private def withSupportingDocsAnswers(
+    answers: UploadSupportingDocuments
+  )(f: Boolean => Future[Result]): Future[Result] =
+    answers
+      .fold(_.hasSupportingDocuments, c => Some(c.hasSupportingDocuments))
+      .fold[Future[Result]](Future.successful(Redirect(routes.UploadSupportingDocumentsController.checkYourAnswers())))(
+        f
+      )
+
+  private def commonDisplayBehaviour[A, P: Writeable, R](
+    currentAnswers: UploadSupportingDocuments
+  )(form: UploadSupportingDocuments => Form[A])(
+    page: (Form[A], Call) => P
+  )(
+    requiredPreviousAnswer: UploadSupportingDocuments => Option[R],
+    redirectToIfNoRequiredPreviousAnswer: Call
+  ): Future[Result] =
+    if (requiredPreviousAnswer(currentAnswers).isDefined) {
+      val backLink = currentAnswers.fold(
+        _ => redirectToIfNoRequiredPreviousAnswer,
+        _ => routes.UploadSupportingDocumentsController.checkYourAnswers()
+      )
+      Future.successful(Ok(page(form(currentAnswers), backLink)))
+    } else {
+      Future.successful(Redirect(redirectToIfNoRequiredPreviousAnswer))
```

Comment:
if you want to avoid writing `Future.successful` everywhere, you can import `utils.toFuture`. That has an implicit conversion from `Result` to `Future[Result` in it

Reply:
changed to use this